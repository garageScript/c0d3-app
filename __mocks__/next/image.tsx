import React from 'react'
import type { ImageProps } from 'next/image'
import _ from 'lodash'

const Image = (props: ImageProps) => {
  /*not all Image properties are allowed in <img/> tag
  ObjectFit causes warnings, but otherwise lessons/error images won't keep their dimensions 
  see https://developer.mozilla.org/en-US/docs/Web/CSS/object-fit*/
  const copy = _.omit(props, 'objectFit')
  return <img {...copy} />
}

export default Image
