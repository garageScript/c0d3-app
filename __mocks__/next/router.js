const push = jest.fn()
const query = {}
export const useRouter = jest.fn().mockImplementation(() => {
  return {
    route: '/',
    pathname: '/',
    query,
    asPath: '/',
    push: push,
    replace: async () => true,
    reload: () => null,
    back: () => null
  }
})
