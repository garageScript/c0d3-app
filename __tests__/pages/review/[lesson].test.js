import React from 'react'
import {
  render,
  waitFor,
  screen,
  waitForElementToBeRemoved
} from '@testing-library/react'
import '@testing-library/jest-dom'
import { MockedProvider } from '@apollo/client/testing'
import GET_APP from '../../../graphql/queries/getApp'
import GET_SUBMISSONS from '../../../graphql/queries/getSubmissions'
import Review from '../../../pages/review/[lesson]'
import dummyLessonData from '../../../__dummy__/lessonData'
import dummySessionData from '../../../__dummy__/sessionData'
import dummyAlertData from '../../../__dummy__/alertData'
import { useRouter } from 'next/router'
import expectLoading from '../../utils/expectLoading'

const getAppMock = {
  request: { query: GET_APP },
  result: {
    data: {
      session: {
        ...dummySessionData,
        lessonStatus: [
          {
            lessonId: 2,
            isPassed: '1614694120099', // this column is a timestamp on the DB
            isTeaching: null,
            isEnrolled: null,
            starGiven: null
          }
        ]
      },
      lessons: dummyLessonData,
      alerts: dummyAlertData
    }
  }
}
const getSubmissionsMock = {
  request: { query: GET_SUBMISSONS, variables: { lessonId: 2 } },
  result: {
    data: {
      submissions: [
        {
          id: 1,
          status: 'open',
          diff:
            'diff --git a/js7/1.js b/js7/1.js\nindex 9c96b34..853bddf 100644\n--- a/js7/1.js\n+++ b/js7/1.js\n@@ -1,8 +1,19 @@\n-// write your code here!\n const solution = () => {\n-  // global clear all timeout:\n+  const allT = [];\n+  const old = setTimeout;\n+  window.setTimeout = (func, delay) => {\n+    const realTimeout = old(func, delay);\n+    allT.push(realTimeout);\n+    return realTimeout;\n+  };\n+  window.clearAllTimouts = () => {\n+    while (allT.length) {\n+      clearTimeout(allT.pop());\n+    }\n+  };\n   cat = () => {\n-  }\n+    window.clearAllTimouts();\n+  };\n };\n \n module.exports = solution;\n',
          comment: 'TEST 2',
          challenge: {
            title: 'Sum of 1 Numbers'
          },
          challengeId: 107,
          lessonId: 5,
          user: {
            id: '6',
            username: 'newbie'
          },
          reviewer: {
            id: '0',
            username: 'admin',
            name: 'Admin Admin'
          },
          createdAt: '',
          updatedAt: ''
        }
      ]
    }
  }
}
const mocks = [getAppMock, getSubmissionsMock]
describe('Lesson Page', () => {
  const { query, push } = useRouter()
  query['lesson'] = '2'
  test('Should render new submissions', async () => {
    const { container } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <Review />
      </MockedProvider>
    )
    await waitFor(() =>
      expect(
        screen.getByRole('heading', { name: 'newbie - Sum of 1 Numbers' })
      ).toBeTruthy()
    )
    await waitFor(() => expect(container).toMatchSnapshot())
  })
  test('Should return loading spinner when loading', () => {
    expectLoading(<Review />)
  })
  test('Should redirect to login if no session', async () => {
    const noSessionMock = {
      request: { query: GET_APP },
      result: {
        data: {
          session: null,
          lessons: dummyLessonData,
          alerts: dummyAlertData
        }
      }
    }
    render(
      <MockedProvider
        mocks={[noSessionMock, getSubmissionsMock]}
        addTypename={false}
      >
        <Review />
      </MockedProvider>
    )

    await waitFor(() => expect(push).toBeCalledWith('/login'))
  })
  test("Should redirect to curriculum if user hasn't completed lesson yet", async () => {
    const noSessionMock = {
      request: { query: GET_APP },
      result: {
        data: {
          session: dummySessionData,
          lessons: dummyLessonData,
          alerts: dummyAlertData
        }
      }
    }
    render(
      <MockedProvider
        mocks={[noSessionMock, getSubmissionsMock]}
        addTypename={false}
      >
        <Review />
      </MockedProvider>
    )
    await waitFor(() => expect(push).toBeCalledWith('/curriculum'))
  })
  test('Should render empty submissions', async () => {
    const noSubmissionMock = { ...getSubmissionsMock, result: null }
    const { container } = render(
      <MockedProvider
        mocks={[getAppMock, noSubmissionMock]}
        addTypename={false}
      >
        <Review />
      </MockedProvider>
    )

    await waitForElementToBeRemoved(() => screen.queryByText('Loading...'))
    await waitFor(() => expect(container).toMatchSnapshot())
  })
  test('Should render Error component if route is invalid', async () => {
    query['lesson'] = '100'
    const { container } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <Review />
      </MockedProvider>
    )
    await waitFor(() => screen.getAllByText(/Page not found/i))

    await waitFor(() => expect(container).toMatchSnapshot())
  })
})
