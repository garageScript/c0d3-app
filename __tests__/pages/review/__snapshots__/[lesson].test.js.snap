// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lesson Page Should render Error component if route is invalid 1`] = `
<div>
  <nav
    class="navbar navbar-expand-lg navbar-light justify-content-between bg-white"
  >
    <div
      class="container"
    >
      <a
        class="navbar-brand text-primary font-weight-bold"
        href="/"
      >
        C0D3
      </a>
      <div
        id="navbarNav"
      >
        <div
          class="navbar-nav collapse navbar-collapse"
        >
          <div
            class="navbar-nav collapse navbar-collapse"
          >
            <a
              class="nav-item nav-link"
              href="/curriculum"
            >
              Curriculum
            </a>
            <a
              class="nav-item nav-link"
              href="https://github.com/garageScript/c0d3-app"
              rel="noopener noreferrer"
              target="_blank"
            >
              Repo
            </a>
            <a
              class="nav-item nav-link"
              href="https://www.notion.so/Table-of-Contents-a83980f81560429faca3821a9af8a5e2"
              rel="noopener noreferrer"
              target="_blank"
            >
              Journey
            </a>
            <a
              class="nav-item nav-link"
              href="https://chat.c0d3.com"
              rel="noopener noreferrer"
              target="_blank"
            >
              Help
            </a>
            <a
              class="nav-item nav-link"
              href="/contributors"
            >
              Contributors
            </a>
            <div
              class="dropdown"
            >
              <button
                aria-expanded="false"
                aria-haspopup="true"
                class="dropdown-toggle btn btn-none"
                type="button"
              >
                Admin
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="d-flex"
      >
        <a
          class="btn btn-secondary border overflow-hidden p-2 text-truncate"
          href="/profile/fakeusername"
        >
           fakeusername
        </a>
        <button
          class="btn border ml-2"
          style="color: rgb(41, 41, 41);"
        >
          Logout
        </button>
      </div>
    </div>
  </nav>
  <div
    class="container"
  >
    <div
      class="container"
    >
      <div
        class="row"
      >
        <div
          class="d-flex col-sm-3 align-items-center justify-content-center"
        >
          <div
            class="text-center mt-3"
          >
            <div
              style="color: rgb(41, 41, 41); font-size: 1.5rem;"
            >
              <strong>
                Page not found
              </strong>
            </div>
            <div
              style="color: rgb(41, 41, 41); font-size: 1rem;"
            >
              Page not found
            </div>
            <a
              class="btn btn-primary py-3 px-5 mt-3"
              href="/"
            >
              <h3
                class="font-weight-bold"
              >
                Back
              </h3>
            </a>
          </div>
        </div>
        <div
          class="mb-3 col-sm-9"
        >
          <img
            height="1200"
            layout="responsive"
            objectfit="contain"
            src="/assets/errors/404.svg"
            width="1200"
          />
        </div>
      </div>
    </div>
  </div>
  <footer
    class="mt-5 font-weight-light text-center pb-5 text-muted"
  >
    © Copyright 2020 GarageScript
    <br />
    All Rights Reserved
  </footer>
</div>
`;

exports[`Lesson Page Should render empty submissions 1`] = `
<div>
  <div>
    <nav
      class="navbar navbar-expand-lg navbar-light justify-content-between bg-white"
    >
      <div
        class="container"
      >
        <a
          class="navbar-brand text-primary font-weight-bold"
          href="/"
        >
          C0D3
        </a>
        <div
          id="navbarNav"
        >
          <div
            class="navbar-nav collapse navbar-collapse"
          >
            <div
              class="navbar-nav collapse navbar-collapse"
            >
              <a
                class="nav-item nav-link"
                href="/curriculum"
              >
                Curriculum
              </a>
              <a
                class="nav-item nav-link"
                href="https://github.com/garageScript/c0d3-app"
                rel="noopener noreferrer"
                target="_blank"
              >
                Repo
              </a>
              <a
                class="nav-item nav-link"
                href="https://www.notion.so/Table-of-Contents-a83980f81560429faca3821a9af8a5e2"
                rel="noopener noreferrer"
                target="_blank"
              >
                Journey
              </a>
              <a
                class="nav-item nav-link"
                href="https://chat.c0d3.com"
                rel="noopener noreferrer"
                target="_blank"
              >
                Help
              </a>
              <a
                class="nav-item nav-link"
                href="/contributors"
              >
                Contributors
              </a>
              <div
                class="dropdown"
              >
                <button
                  aria-expanded="false"
                  aria-haspopup="true"
                  class="dropdown-toggle btn btn-none"
                  type="button"
                >
                  Admin
                </button>
              </div>
            </div>
          </div>
        </div>
        <div
          class="d-flex"
        >
          <a
            class="btn btn-secondary border overflow-hidden p-2 text-truncate"
            href="/profile/fakeusername"
          >
             fakeusername
          </a>
          <button
            class="btn border ml-2"
            style="color: rgb(41, 41, 41);"
          >
            Logout
          </button>
        </div>
      </div>
    </nav>
    <div
      class="container"
    >
      <div
        class="row mt-4"
      >
        <div
          class="card shadow-sm mt-3 col-12 px-0 pt-3 border-0"
        >
          <div
            class="card-body p-0"
          >
            <div
              class="d-flex mb-3 px-3"
            >
              <img
                alt="lesson-cover"
                class="lessonTitleCard__lesson-cover mr-3"
                src="/assets/curriculum/js-1-cover.svg"
              />
              <div>
                <p
                  class="m-0"
                >
                  <a
                    class=""
                    href="/curriculum"
                  >
                    Go Back
                  </a>
                </p>
                <h1
                  class="lessonTitleCard__lesson-title"
                >
                  Variables & Functions
                </h1>
              </div>
            </div>
            <div
              class="card-footer bg-white p-0"
            >
              <a
                class="btn border-right rounded-0 px-4 py-3"
                href="https://www.notion.so/garagescript/JS-1-Functions-01dd8400b85f40d083966908acbfa184"
                rel="noopener noreferrer"
                target="_blank"
              >
                LESSON
              </a>
              <a
                class="btn border-right rounded-0 px-4 py-3"
                href="/curriculum/2"
              >
                CHALLENGES
              </a>
              <a
                class="btn border-right rounded-0 px-4 py-3"
                href="/review/2"
              >
                REVIEW
              </a>
            </div>
          </div>
        </div>
        <div
          class="submissions-container container p-0"
        />
      </div>
    </div>
    <footer
      class="mt-5 font-weight-light text-center pb-5 text-muted"
    >
      © Copyright 2020 GarageScript
      <br />
      All Rights Reserved
    </footer>
  </div>
</div>
`;

exports[`Lesson Page Should render new submissions 1`] = `
<div>
  <div
    class="container-fluid bg-primary d-flex flex-column justify-content-center align-items-center"
    style="min-height: 100vh; position: fixed; top: 0px; left: 0px; z-index: 1;"
  >
    <h1
      class="text-white font-weight-bold display-3"
      style="font-family: 'PT Mono', sans-serif;"
    >
      C0D3
    </h1>
    <div
      class="spinner-border text-white"
      role="status"
    >
      <h1
        class="sr-only"
      >
        Loading...
      </h1>
    </div>
  </div>
</div>
`;
